language/bytecode-verifier/src/binary_views.rs

```diff
@@ -0,0 +1,273 @@
+        }
+    }
+
+    pub(crate) fn function_handle_at(&self, idx: FunctionHandleIndex) -> &FunctionHandle {
+        match self {
+            BinaryIndexedView::Module(module) => module.function_handle_at(idx),
+            BinaryIndexedView::Script(script) => script.function_handle_at(idx),
+        }
+    }
+
+    pub(crate) fn function_instantiation_at(
+        &self,
+        idx: FunctionInstantiationIndex,
+    ) -> &FunctionInstantiation {
+        match self {
+            BinaryIndexedView::Module(module) => module.function_instantiation_at(idx),
+            BinaryIndexedView::Script(script) => script.function_instantiation_at(idx),
+        }
+    }
+
+    pub(crate) fn field_handle_at(&self, idx: FieldHandleIndex) -> VMResult<&FieldHandle> {
+        match self {
+            BinaryIndexedView::Module(module) => Ok(module.field_handle_at(idx)),
+            BinaryIndexedView::Script(_) => {
+                Err(VMStatus::new(StatusCode::INVALID_OPERATION_IN_SCRIPT))
+            }
+        }
+    }
+
+    pub(crate) fn struct_instantiation_at(
+        &self,
+        idx: StructDefInstantiationIndex,
+    ) -> VMResult<&StructDefInstantiation> {
+        match self {
+            BinaryIndexedView::Module(module) => Ok(module.struct_instantiation_at(idx)),
+            BinaryIndexedView::Script(_) => {
+                Err(VMStatus::new(StatusCode::INVALID_OPERATION_IN_SCRIPT))
+            }
+        }
+    }
+
+    pub(crate) fn field_instantiation_at(
+        &self,
+        idx: FieldInstantiationIndex,
+    ) -> VMResult<&FieldInstantiation> {
+        match self {
+            BinaryIndexedView::Module(module) => Ok(module.field_instantiation_at(idx)),
+            BinaryIndexedView::Script(_) => {
+                Err(VMStatus::new(StatusCode::INVALID_OPERATION_IN_SCRIPT))
+            }
+        }
+    }
+
+    pub(crate) fn struct_def_at(&self, idx: StructDefinitionIndex) -> VMResult<&StructDefinition> {
+        match self {
+            BinaryIndexedView::Module(module) => Ok(module.struct_def_at(idx)),
+            BinaryIndexedView::Script(_) => {
+                Err(VMStatus::new(StatusCode::INVALID_OPERATION_IN_SCRIPT))
+            }
+        }
+    }
+
+    pub(crate) fn function_def_at(
+        &self,
+        idx: FunctionDefinitionIndex,
+    ) -> VMResult<&FunctionDefinition> {
+        match self {
+            BinaryIndexedView::Module(module) => Ok(module.function_def_at(idx)),
+            BinaryIndexedView::Script(_) => {
+                Err(VMStatus::new(StatusCode::INVALID_OPERATION_IN_SCRIPT))
+            }
+        }
+    }
+
+    // Return the `Kind` of a `SignatureToken` given a context.
+    // A `TypeParameter` used the kind of the `constraints`.
+    // `StructInstantiation` have to "merge" the kinds of the instantiation with that
+    // of the generic type.
+    pub(crate) fn kind(&self, ty: &SignatureToken, constraints: &[Kind]) -> Kind {
```

Comment:
Has this code finally found a forever home...? I feel like this has been moved in everyone of your past few PRs I've reviewed :p

Reply:
LOL - it's so true!!!! 
I am actually liking this location, thoughts?

Reply:
I think its better than anywhere else I've seen so far

Reply:
This also seems good to me. You'll always need an environment if you want to query the kind of a type so it makes sense to me to associate this with `BinaryIndexedView` (the env).